packages/sanity/src/_internal/cli/actions/deploy/helpers.ts

```diff
@@ -0,0 +1,230 @@
+  body: {title?: string; isDefault?: boolean; appHost: string; urlType: 'internal' | 'external'},
+): Promise<UserApplication> {
+  return client.request({uri: '/user-applications', method: 'POST', body})
+}
+
+export interface GetOrCreateUserApplicationOptions {
+  client: SanityClient
+  context: Pick<CliCommandContext, 'output' | 'prompt'>
+  cliConfig?: Pick<CliConfig, 'studioHost'>
+}
+
+export async function getOrCreateUserApplication({
+  client,
+  cliConfig,
+  context: {output, prompt},
+}: GetOrCreateUserApplicationOptions): Promise<UserApplication> {
+  // if there is already an existing user-app, then just return it
+  const existingUserApplication = await getUserApplication({client, appHost: cliConfig?.studioHost})
+  if (existingUserApplication) return existingUserApplication
+
+  // otherwise, we need to create one.
+  // if a `studioHost` was provided in the CLI config, then use that
+  if (cliConfig?.studioHost) {
+    return await createUserApplication(client, {
+      appHost: cliConfig.studioHost,
+      urlType: 'internal',
+    })
+  }
+
+  // otherwise, prompt the user for a hostname
+  output.print('Your project has not been assigned a studio hostname.')
+  output.print('To deploy your Sanity Studio to our hosted Sanity.Studio service,')
+  output.print('you will need one. Please enter the part you want to use.')
+
+  const {promise, resolve} = promiseWithResolvers<UserApplication>()
+
+  await prompt.single({
+    type: 'input',
+    filter: (inp: string) => inp.replace(/\.sanity\.studio$/i, ''),
+    message: 'Studio hostname (<value>.sanity.studio):',
+    // if a string is returned here, it is relayed to the user and prompt allows
+    // the user to try again until this function returns true
+    validate: async (appHost: string) => {
+      try {
+        const response = await createUserApplication(client, {
+          appHost,
+          isDefault: true,
+          urlType: 'internal',
+        })
+        resolve(response)
+        return true
+      } catch (e) {
+        // if the name is taken, it should return a 409 so we relay to the user
+        if (e?.statusCode === 409) return e?.message || 'Conflict' // just in case
+
+        // otherwise, it's a fatal error
+        throw e
+      }
+    },
+  })
+
+  return await promise
+}
+
+export interface CreateDeploymentOptions {
+  client: SanityClient
+  applicationId: string
+  version: string
+  isAutoUpdating: boolean
+  tarball: Gzip
+}
+
+export async function createDeployment({
+  client,
+  tarball,
+  applicationId,
+  isAutoUpdating,
+  version,
+}: CreateDeploymentOptions): Promise<void> {
+  const formData = new FormData()
+  formData.append('isAutoUpdating', isAutoUpdating.toString())
+  formData.append('version', version)
+  formData.append('tarball', tarball, {contentType: 'application/gzip', filename: 'app.tar.gz'})
+
+  await client.request({
+    uri: `/user-applications/${applicationId}/deployments`,
+    method: 'POST',
```

Comment:
❓ 🤔 would this need `headers: { ...formData.getHeaders() }` to get the correct multipart boundary?